cli/src/formatter.rs

```diff
@@ -641,18 +658,33 @@ impl Write for FormatRecorder {
     }
 }
 
+struct RawTextRecorder<'a>(&'a mut FormatRecorder);
+
+impl<'a> Write for RawTextRecorder<'a> {
+    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
+        self.0.push_op(FormatOp::PushRawText);
+        let result = self.0.write(data);
+        self.0.push_op(FormatOp::PopRawText);
+        result
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.0.flush()
+    }
+}
+
 impl Formatter for FormatRecorder {
     fn raw(&mut self) -> &mut dyn Write {
-        panic!("raw output isn't supported by FormatRecorder")
+        Box::leak(Box::new(RawTextRecorder(self)))
```

Comment:
So it leaks memory.

There are a couple of ideas to address this:
1. make it return `Box<dyn Write>` (which seems okay as `.raw()` isn't called in hot loop.)
2. make it return `Option<_>` or `io::Result<_>`, and replace panic with `None`/`Err(unsupported)`.
3. add `.write_raw()` (or enter/exit) formatter method.

I think (2) is the easiest option if you don't need `raw_text()` within reformatting template.

Somewhat related to this, `ColorFormatter::raw()` might have to flush escape codes (i.e. call `write_new_style()`.)

Reply:
Is the leak here really a problem? Unless we have a `jj watch` command at some point, it seems okay to me to leak a few `RawTextRecorder`s for the few seconds jj runs?

I guess it might be a problem if someone is looking at a large repo with jj log or something, seems like an edge case though?

Was just curious what you think, looking into possible fixes in the meantime (thanks for the ideas!).

Reply:
It wouldn't be a real problem, but is a bad design. I would say this is a blocker.